commands/translate.py

```diff
@@ -0,0 +1,169 @@
+from commands.base import Command
+from discord import Embed
+from helpers import *
+from utils.embed_table import EmbedTable
+
+import googletrans
+import json
+import re
+
+HISTORY = 10
+HISTORY_LIMIT = 100
+FLAGS_FILE = "files/flags.json"
+AVATAR_FORMAT = "https://cdn.discordapp.com/avatars/{0.id}/{0.avatar}.png?size=128"
+TRANSLATE_COLOR = int('4885ed', 16)
+
+class Translate(Command):
+    desc = """Translates a marked message to another language (removes all emojis).
+    React with a flag and type `!translate` to translate the message to the corresponding language.
+    Optional `index` argument defines offset of message to translate in list of marked messages."""
+
+    async def eval(self, index=0):
+        if int(index) < 0:
+            raise CommandFailure('Invalid index!')
+
+        archive = await create_list(self.client.logs_from(self.message.channel,
+                                                            limit=HISTORY_LIMIT))
+
+        if not archive:
+            raise CommandFailure('No tagged messages in the last %s found!' % HISTORY_LIMIT)
+
+        if int(index) >= len(archive):
+            raise CommandFailure('Invalid index!')
+
+        lang_list = archive[int(index)]
+        for entry in lang_list:
+            #translate contents
+            translator = googletrans.Translator()
+            text = removeEmojis(entry.content)
+            entry.content = translator.translate(text, dest=entry.language).text
+
+            channel = self.message.channel
+            header = ''
+            if type(entry.flag) is str:
+                footer = "Translation to %s requested by %s" % (entry.flag, self.name)
+            else:
+                footer = "Translation requested by %s" % self.name
+            await self.client.send_message(channel, header, embed=entry.as_embed(footer))
+
+
+class Add(Translate):
+    desc = "Adds mapping between an emoji and a Google API language code. Mods only."
+    roles_required = ['mod', 'exec']
+
+    def eval(self, emoji, langcode):
+        #check if valid language code
+        if langcode not in googletrans.LANGUAGES:
+            raise CommandFailure("Language code not valid!")
+
+        # Open the JSON file or create a new dict to load
+        try:
+            with open(FLAGS_FILE, 'r') as old:
+                flags = json.load(old)
+        except FileNotFoundError:
+            flags = {}
+
+        flags[emoji] = langcode
```

Comment:
It's also possible for old mappings to be overridden (eg if I map flag_au --> en, I can then do `!translate add flag_au fi`). I'd like there to be a message that says "mapping for flag_au already exists, please remove it first" in case previous mappings are overridden.